### ISSUE TITLE
Consider removing corefx repo build dependency on libcurl

### ISSUE THREAD

**Issue by ghost** [dotnet/corefx#28353]
(https://github.com/dotnet/corefx/issues/27783#issuecomment-375096678)

For build (everything) [from source](https://github.com/dotnet/source-build), dependency on libcurl can be made optional, such that if the project was compiled in the absence of libcurl, this line instead of throwing fatal error:

https://github.com/dotnet/corefx/blob/9a2e5d8f0e7f816f983bce75f5bba22811df4994/src/Native/Unix/System.Net.Http.Native/CMakeLists.txt#L4-L6 

will show a message. Then the `PAL_` methods that deal with curl will be `#ifdef`'d out. On managed side, unsetting `DOTNET_SYSTEM_NET_HTTP_USESOCKETSHTTPHANDLE` or setting it to `false` will continue to use SocketHttpSocketHandler and emit (per-initialization?) EventSource warning log (https://github.com/dotnet/corefx/issues/2500) to notify "CurlHandler is not present in this build, continue using SocketHttpSocketHandler ".

**Comment by Petermarcu** [comment]
Good suggestion

**Comment by omajid** [comment]
I would ask for the opposite, though: please do not carry around another copy of a well known library.

It's really good to be able to build .NET Core from source. In fact, I would say something is not open source if it can't be build.

But there is a tangential problem of bundling: the more copies of a library exist on a system, the harder it is to fix them if a problem comes up. Common libraries like curl (and libunwind) make this worse. Consider that curl gets a security vulnerability every few weeks. Copying that code means there is additional burden to .NET Core for maintaining the fork (and avoiding any divergences). For anyone who finds out about a security issue, they now need to find all embedded copies of libraries and fix those too. IMO, we shouldn't copy libraries that are much easier to build and use separately. Creating a separate copy of curl just means there is another (likely vulnerable to security exploits) copy of curl floating on your system, without any particular advantages.

If you can't get curl on a system, scripts to build them would be nice, and it would be especially nice to upstream that into the curl project proper so there is one well known place to get everything for anyone looking to get a version of curl they need.

**Comment by Petermarcu** [comment]
We are not going to bundle libcurl. Your points are extremely valid (especially for libcurl). We have done work in 2.1 to not need libcurl anymore by default. You don't need it in your OS or in your app. We have a full .NET based client networking stack. This issue is about reacting to the fact that its not necessary. Its not to bundle it.

**Comment by kasper3** [comment]
Starting from .NET Core 2.1, `System.Net.Http.SocketsHttpHandler` is default which does not require libccurl. The CurlHandler is opt-in using `DOTNET_SYSTEM_NET_HTTP_USESOCKETSHTTPHANDLE` environment variable: https://github.com/dotnet/corefx/blob/d7548549/src/System.Net.Http/src/System/Net/Http/HttpClientHandler.cs, (inverted in 2b850f6).

With proposed approach, by allowing corefx to compile without libcurl present, setting the environment variable  `DOTNET_SYSTEM_NET_HTTP_USESOCKETSHTTPHANDLE`  to `false` will continue using `SocketsHttpHandler` and emit a warning to event source at runtime.

**Comment by wfurt** [comment]
Even if not default, the curl handler is part of the product. Also, you should probably not mix runtime and build time settings. This should probably be moved to source-build to make build more flexible. 
Once can possibly argue this for there dependencies like kerberos or ssl. I would expect practical impact only one very small embedded system @kasper3 .

**Comment by kasper3** [comment]
Maybe make libcurl dependency same as libssl using shim https://github.com/dotnet/corefx/blob/051f53b874ad7eb16c817d72c7ee8aa5a4fa1c79/src/Native/Unix/System.Security.Cryptography.Native/opensslshim.h. Though the build machine still needs to have libcurl (like in case of openssl) to <ins>build</ins>, the target system will continue to <ins>work</ins> on runtime.

**Comment by stephentoub** [comment]
> Though the build machine still needs to have libcurl (like in case of openssl) to build, the target system will continue to work on runtime.

That is the plan.

**Comment by kasper3** [comment]
`dotnet new` fails without libcurl even when CurlHandle is disabled.

With new Ubuntu minimal installation (e.g. `docker run -it ubuntu`), installed dotnet using generic linux tarball

```sh
apt update
apt install wget libunwind-dev libicu-dev
wget https://dotnetcli.azureedge.net/dotnet/Sdk/2.1.300-preview3-008414/dotnet-sdk-2.1.300-preview3-008414-linux-x64.tar.gz
mkdir -p /usr/bin/dotnet
tar -xzvf dotnet-sdk-2.1.300-preview3-008414-linux-x64.tar.gz  -C /usr/bin/dotnet
export PATH=$PATH:/usr/bin/dotnet
rm dotnet-sdk-2.1.300-preview3-008414-linux-x64.tar.gz
```

Then running `dotnet new console -n test` fails with:

```sh
Running 'dotnet restore' on test/test.csproj...
  Restoring packages for /test/test.csproj...
/usr/bin/dotnet/sdk/2.1.300-preview3-008414/NuGet.targets(114,5): error : Unable to load the service index for source https://api.nuget.org/v3/index.json. [/test/test.csproj]
/usr/bin/dotnet/sdk/2.1.300-preview3-008414/NuGet.targets(114,5): error :   The type initializer for 'System.Net.Http.CurlHandler' threw an exception. [/test/test.csproj]
/usr/bin/dotnet/sdk/2.1.300-preview3-008414/NuGet.targets(114,5): error :   The type initializer for 'Http' threw an exception. [/test/test.csproj]
/usr/bin/dotnet/sdk/2.1.300-preview3-008414/NuGet.targets(114,5): error :   The type initializer for 'HttpInitializer' threw an exception. [/test/test.csproj]
/usr/bin/dotnet/sdk/2.1.300-preview3-008414/NuGet.targets(114,5): error :   Unable to load shared library 'System.Net.Http.Native' or one of its dependencies. In order to help diagnose loading problems, consider setting the LD_DEBUG environment variable: libSystem.Net.Http.Native: cannot open shared object file: No such file or directory [/test/test.csproj]

Restore failed.
Post action failed.
```

Perhaps changing the libcurl binding in a lazy manner (shim/dlopen approach used in `opensslshim.cpp`) would prevent it from requiring libcurl eagerly?

**Comment by Petermarcu** [comment]
I don't think there has been a complete end to end build of the SDK with the curlhander off by default. I've been waiting for one. Did you try setting the env var?

**Comment by kasper3** [comment]
Ah that change was made 11 days ago, thought CLI must have picked up.
Tried with
`rm -rf test; DOTNET_SYSTEM_NET_HTTP_USESOCKETSHTTPHANDLE=true dotnet new console -n test`
and (with old environment name https://github.com/dotnet/corefx/pull/26763)
`rm -rf test; COMPlus_UseManagedHttpClientHandler=true dotnet new console -n test`
same error (unless I run `apt install libcurl4-openssl-dev`)

**Comment by Petermarcu** [comment]
There are changes happening to the build system so the flow is not ideal right now. It could have made it through though. I do see merges into the cli repo that say it should have picked up the changes.

**Comment by Petermarcu** [comment]
Great to have confirmation that you don't see the runtime dependency. My understanding is that we are planning to use 2.0 to build 2.1 and then 2.1 to build 2.2. I think that means we don't have a great path to removing the build time dependency in 2.1.

**Comment by karelz** [comment]
Will be solved by libcurl removal entirely in 5.0: #41183
